# Gesture Recognition
In this group project, you are going to build a 3D Conv model that will be able to predict the 5 gestures correctly. Please import the following libraries to get started.

In [1]:
import numpy as np
import os
from scipy.misc import imread, imresize
import datetime
import os
import warnings
warnings.filterwarnings("ignore")
import abc
from sys import getsizeof
import shutil
import abc


In [2]:
import glob, os , shutil
for f in glob.glob("model_init*"):
    shutil.rmtree(f)

We set the random seed so that the results don't vary drastically.

In [3]:
np.random.seed(30)
import random as rn
rn.seed(30)
import tensorflow as tf
tf.random.set_seed(30)

**In this block, you read the folder names for training and validation. You also set the batch_size here. Note that you set the batch size in such a way that you are able to use the GPU in full capacity. You keep increasing the batch size until the machine throws an error.**

In [4]:
from keras.models import Sequential, Model
from keras.layers import Dense, GRU, Flatten, TimeDistributed, Flatten, BatchNormalization, Activation
from keras.layers.convolutional import Conv3D, MaxPooling3D, Conv2D, MaxPooling2D
from keras.layers.recurrent import LSTM
from keras.callbacks import ModelCheckpoint, ReduceLROnPlateau
from keras import optimizers
from keras.layers import Dropout

In [5]:
# the entire dataset is placed in below directory
main_folder='Project_data'

### Plotting

In [6]:
def plot(Model):
    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(15,4))
    axes[0].plot(Model.history['loss'])   
    axes[0].plot(Model.history['val_loss'])
    axes[0].legend(['loss','val_loss'])

    axes[1].plot(Model.history['categorical_accuracy'])   
    axes[1].plot(Model.history['val_categorical_accuracy'])
    axes[1].legend(['categorical_accuracy','val_categorical_accuracy'])

## Generator
This is one of the most important part of the code. The overall structure of the generator has been given. In the generator, you are going to preprocess the images as you have images of 2 different dimensions as well as create a batch of video frames. You have to experiment with `img_idx`, `y`,`z` and normalization such that you get high accuracy.

In [7]:
# referred ABC library use case from this link https://riptutorial.com/python/example/23083/why-how-to-use-abcmeta-and--abstractmethod

class ModelBuilder(metaclass= abc.ABCMeta):
    
    def initialize_src_path(self,main_folder):
        self.train_doc = np.random.permutation(open(main_folder + '/' + 'train.csv').readlines())
        self.val_doc = np.random.permutation(open(main_folder + '/' + 'val.csv').readlines())
        self.train_path = main_folder + '/' + 'train'
        self.val_path =  main_folder + '/' + 'val'
        self.num_train_sequences = len(self.train_doc)
        self.num_val_sequences = len(self.val_doc)
        
    def initialize_image_properties(self,image_height=100,image_width=100):
        self.image_height=image_height
        self.image_width=image_width
        self.channels=3
        self.num_classes=5
        self.total_frames=30
          
    def initialize_hyperparams(self,frames_to_sample=30,batch_size=40,num_epochs=20):
        self.frames_to_sample=frames_to_sample
        self.batch_size=batch_size
        self.num_epochs=num_epochs
        
        
    def generator(self,source_path, folder_list, augment=False):
        img_idx = np.round(np.linspace(0,self.total_frames-1,self.frames_to_sample)).astype(int)
        batch_size=self.batch_size
        while True:
            t = np.random.permutation(folder_list)
            num_batches = len(t)//batch_size
        
            for batch in range(num_batches): 
                batch_data, batch_labels= self.one_batch_data(source_path,t,batch,batch_size,img_idx,augment)
                yield batch_data, batch_labels 

            remaining_seq=len(t)%batch_size
        
            if (remaining_seq != 0):
                batch_data, batch_labels= self.one_batch_data(source_path,t,num_batches,batch_size,img_idx,augment,remaining_seq)
                yield batch_data, batch_labels 
    
    
    def one_batch_data(self,source_path,t,batch,batch_size,img_idx,augment,remaining_seq=0):
    
        seq_len = remaining_seq if remaining_seq else batch_size
    
        batch_data = np.zeros((seq_len,len(img_idx),self.image_height,self.image_width,self.channels)) 
        batch_labels = np.zeros((seq_len,self.num_classes)) 
    
        if (augment): batch_data_aug = np.zeros((seq_len,len(img_idx),self.image_height,self.image_width,self.channels))

        
        for folder in range(seq_len): 
            imgs = os.listdir(source_path+'/'+ t[folder + (batch*batch_size)].split(';')[0]) 
            for idx,item in enumerate(img_idx): 
                image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
                image_resized=imresize(image,(self.image_height,self.image_width,3))
            

                batch_data[folder,idx,:,:,0] = (image_resized[:,:,0])/255
                batch_data[folder,idx,:,:,1] = (image_resized[:,:,1])/255
                batch_data[folder,idx,:,:,2] = (image_resized[:,:,2])/255
            
                if (augment):
                    shifted = cv2.warpAffine(image, 
                                             np.float32([[1, 0, np.random.randint(-30,30)],[0, 1, np.random.randint(-30,30)]]), 
                                            (image.shape[1], image.shape[0]))
                    
                    gray = cv2.cvtColor(shifted,cv2.COLOR_BGR2GRAY)

                    x0, y0 = np.argwhere(gray > 0).min(axis=0)
                    x1, y1 = np.argwhere(gray > 0).max(axis=0) 
                    
                    cropped=shifted[x0:x1,y0:y1,:]
                    
                    image_resized=imresize(cropped,(self.image_height,self.image_width,3))
            
                    batch_data_aug[folder,idx,:,:,0] = (image_resized[:,:,0])/255
                    batch_data_aug[folder,idx,:,:,1] = (image_resized[:,:,1])/255
                    batch_data_aug[folder,idx,:,:,2] = (image_resized[:,:,2])/255
                
            
            batch_labels[folder, int(t[folder + (batch*batch_size)].strip().split(';')[2])] = 1
            
    
        if (augment):
            batch_data=np.concatenate([batch_data,batch_data_aug])
            batch_labels=np.concatenate([batch_labels,batch_labels])

        
        return(batch_data,batch_labels)
    
    
    def train_model(self, model, augment_data=False):
        train_generator = self.generator(self.train_path, self.train_doc,augment=augment_data)
        val_generator = self.generator(self.val_path, self.val_doc)

        model_name = 'model_init' + '_' + str(datetime.datetime.now()).replace(' ','').replace(':','_') + '/'
    
        if not os.path.exists(model_name):
            os.mkdir(model_name)
        
        filepath = model_name + 'model-{epoch:05d}-{loss:.5f}-{categorical_accuracy:.5f}-{val_loss:.5f}-{val_categorical_accuracy:.5f}.h5'

        checkpoint = ModelCheckpoint(filepath, monitor='val_loss', verbose=1, save_best_only=False, save_weights_only=False, mode='auto', period=1)
        LR = ReduceLROnPlateau(monitor='val_loss', factor=0.2, verbose=1, patience=4)
        callbacks_list = [checkpoint, LR]

        if (self.num_train_sequences%self.batch_size) == 0:
            steps_per_epoch = int(self.num_train_sequences/self.batch_size)
        else:
            steps_per_epoch = (self.num_train_sequences//self.batch_size) + 1

        if (self.num_val_sequences%self.batch_size) == 0:
            validation_steps = int(self.num_val_sequences/self.batch_size)
        else:
            validation_steps = (self.num_val_sequences//self.batch_size) + 1
    
        history=model.fit_generator(train_generator, steps_per_epoch=steps_per_epoch, epochs=self.num_epochs, verbose=1, 
                            callbacks=callbacks_list, validation_data=val_generator, 
                            validation_steps=validation_steps, class_weight=None, workers=1, initial_epoch=0)
        return history

        
    @abc.abstractmethod
    def define_model(self):
        pass

## Sample Model
Here you make the model using different functionalities that Keras provides. Remember to use `Conv3D` and `MaxPooling3D` and not `Conv2D` and `Maxpooling2D` for a 3D convolution model. You would want to use `TimeDistributed` while building a Conv2D + RNN model. Also remember that the last layer is the softmax. Design the network in such a way that the model is able to give good accuracy on the least number of parameters so that it can fit in the memory of the webcam.

In [8]:
class ModelConv3D1(ModelBuilder):
    
    def define_model(self):

        model = Sequential()
        model.add(Conv3D(16, (3, 3, 3), padding='same',
                 input_shape=(self.frames_to_sample,self.image_height,self.image_width,self.channels)))
        model.add(Activation('relu'))
        model.add(BatchNormalization())
        model.add(MaxPooling3D(pool_size=(2, 2, 2)))

        model.add(Conv3D(32, (2, 2, 2), padding='same'))
        model.add(Activation('relu'))
        model.add(BatchNormalization())
        model.add(MaxPooling3D(pool_size=(2, 2, 2)))

        model.add(Conv3D(64, (2, 2, 2), padding='same'))
        model.add(Activation('relu'))
        model.add(BatchNormalization())
        model.add(MaxPooling3D(pool_size=(2, 2, 2)))

        model.add(Conv3D(128, (2, 2, 2), padding='same'))
        model.add(Activation('relu'))
        model.add(BatchNormalization())
        model.add(MaxPooling3D(pool_size=(2, 2, 2)))

        model.add(Flatten())
        model.add(Dense(128,activation='relu'))
        model.add(BatchNormalization())
        model.add(Dropout(0.5))

        model.add(Dense(64,activation='relu'))
        model.add(BatchNormalization())
        model.add(Dropout(0.25))


        model.add(Dense(self.num_classes,activation='softmax'))

        optimiser = optimizers.Adam()
        #optimiser = 'sgd'
        model.compile(optimizer=optimiser, loss='categorical_crossentropy', metrics=['categorical_accuracy'])
        return model

In [9]:
Conv3D1=ModelConv3D1()
Conv3D1.initialize_src_path(main_folder)
Conv3D1.initialize_image_properties(image_height=160,image_width=160)
Conv3D1.initialize_hyperparams(frames_to_sample=30,batch_size=30,num_epochs=1)
Conv3D1_model=Conv3D1.define_model()
Conv3D1_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv3d (Conv3D)              (None, 30, 160, 160, 16)  1312      
_________________________________________________________________
activation (Activation)      (None, 30, 160, 160, 16)  0         
_________________________________________________________________
batch_normalization (BatchNo (None, 30, 160, 160, 16)  64        
_________________________________________________________________
max_pooling3d (MaxPooling3D) (None, 15, 80, 80, 16)    0         
_________________________________________________________________
conv3d_1 (Conv3D)            (None, 15, 80, 80, 32)    4128      
_________________________________________________________________
activation_1 (Activation)    (None, 15, 80, 80, 32)    0         
_________________________________________________________________
batch_normalization_1 (Batch (None, 15, 80, 80, 32)    1

In [10]:
Conv3D1.train_model(Conv3D1_model)

W0315 11:25:08.159296 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.
W0315 11:25:08.161027 140545046239040 deprecation.py:323] From <ipython-input-7-c85facc09113>:122: Model.fit_generator (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
Instructions for updating:
Please use Model.fit, which supports generators.


23/23 [==============================] - ETA: 0s - loss: 1.5901 - categorical_accuracy: 0.4284
Epoch 00001: saving model to model_init_2021-03-1511_25_08.158874/model-00001-1.59008-0.42836-2.63335-0.21000.h5
23/23 [==============================] - 204s 9s/step - loss: 1.5901 - categorical_accuracy: 0.4284 - val_loss: 2.6334 - val_categorical_accuracy: 0.2100 - lr: 0.0010


In [11]:
Conv3D1_bs40=ModelConv3D1()
Conv3D1_bs40.initialize_src_path(main_folder)
Conv3D1_bs40.initialize_image_properties(image_height=160,image_width=160)
Conv3D1_bs40.initialize_hyperparams(frames_to_sample=30,batch_size=40,num_epochs=1)
Conv3D1_model_bs40=Conv3D1_bs40.define_model()
Conv3D1_model_bs40.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv3d_4 (Conv3D)            (None, 30, 160, 160, 16)  1312      
_________________________________________________________________
activation_4 (Activation)    (None, 30, 160, 160, 16)  0         
_________________________________________________________________
batch_normalization_6 (Batch (None, 30, 160, 160, 16)  64        
_________________________________________________________________
max_pooling3d_4 (MaxPooling3 (None, 15, 80, 80, 16)    0         
_________________________________________________________________
conv3d_5 (Conv3D)            (None, 15, 80, 80, 32)    4128      
_________________________________________________________________
activation_5 (Activation)    (None, 15, 80, 80, 32)    0         
_________________________________________________________________
batch_normalization_7 (Batch (None, 15, 80, 80, 32)   

#### Got below memory exhaust error with image resolution of 160x160, 30 frames and a batch_size of 40
ResourceExhaustedError:  OOM when allocating tensor with shape[40,16,15,80,80] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[node gradient_tape/sequential_2/max_pooling3d_8/MaxPool3D/MaxPool3DGrad (defined at <ipython-input-11-c85facc09113>:122) ]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
 [Op:__inference_train_function_7489]

In [12]:
print("Memory util is {} Gigs". format(getsizeof(np.zeros((40,16,30,160,160)))/(1024*1024*1024)))

Memory util is 3.662109524011612 Gigs


In [13]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=160,image_width=160)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=30,num_epochs=3)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:28:59.951972 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 1736389
Epoch 1/3
23/23 [==============================] - ETA: 0s - loss: 1.6057 - categorical_accuracy: 0.4133
Epoch 00001: saving model to model_init_2021-03-1511_28_59.951207/model-00001-1.60569-0.41327-2.13461-0.21000.h5
23/23 [==============================] - 58s 3s/step - loss: 1.6057 - categorical_accuracy: 0.4133 - val_loss: 2.1346 - val_categorical_accuracy: 0.2100 - lr: 0.0010
Epoch 2/3
23/23 [==============================] - ETA: 0s - loss: 1.1257 - categorical_accuracy: 0.5671
Epoch 00002: saving model to model_init_2021-03-1511_28_59.951207/model-00002-1.12573-0.56712-4.76754-0.18000.h5
23/23 [==============================] - 58s 3s/step - loss: 1.1257 - categorical_accuracy: 0.5671 - val_loss: 4.7675 - val_categorical_accuracy: 0.1800 - lr: 0.0010
Epoch 3/3
23/23 [==============================] - ETA: 0s - loss: 0.8557 - categorical_accuracy: 0.6621
Epoch 00003: saving model to model_init_2021-03-1511_28_59.951207/model-00003-0.85571-0.66214-6.12139-0.1

In [14]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=30,batch_size=30,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:32:02.970877 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
23/23 [==============================] - ETA: 0s - loss: 1.7915 - categorical_accuracy: 0.3560
Epoch 00001: saving model to model_init_2021-03-1511_32_02.970349/model-00001-1.79150-0.35596-2.34787-0.21000.h5
23/23 [==============================] - 90s 4s/step - loss: 1.7915 - categorical_accuracy: 0.3560 - val_loss: 2.3479 - val_categorical_accuracy: 0.2100 - lr: 0.0010
Epoch 2/2
23/23 [==============================] - ETA: 0s - loss: 1.1892 - categorical_accuracy: 0.5370
Epoch 00002: saving model to model_init_2021-03-1511_32_02.970349/model-00002-1.18918-0.53695-4.51897-0.16000.h5
23/23 [==============================] - 91s 4s/step - loss: 1.1892 - categorical_accuracy: 0.5370 - val_loss: 4.5190 - val_categorical_accuracy: 0.1600 - lr: 0.0010


In [15]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=30,batch_size=60,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:35:12.400348 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
12/12 [==============================] - ETA: 0s - loss: 1.5725 - categorical_accuracy: 0.3816
Epoch 00001: saving model to model_init_2021-03-1511_35_12.399726/model-00001-1.57246-0.38160-1.59263-0.22000.h5
12/12 [==============================] - 90s 8s/step - loss: 1.5725 - categorical_accuracy: 0.3816 - val_loss: 1.5926 - val_categorical_accuracy: 0.2200 - lr: 0.0010
Epoch 2/2
12/12 [==============================] - ETA: 0s - loss: 1.2172 - categorical_accuracy: 0.5309
Epoch 00002: saving model to model_init_2021-03-1511_35_12.399726/model-00002-1.21721-0.53092-2.02645-0.16000.h5
12/12 [==============================] - 89s 7s/step - loss: 1.2172 - categorical_accuracy: 0.5309 - val_loss: 2.0265 - val_categorical_accuracy: 0.1600 - lr: 0.0010


In [16]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=60,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:38:25.032883 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
12/12 [==============================] - ETA: 0s - loss: 1.8245 - categorical_accuracy: 0.3499
Epoch 00001: saving model to model_init_2021-03-1511_38_25.032126/model-00001-1.82449-0.34992-1.73380-0.16000.h5
12/12 [==============================] - 49s 4s/step - loss: 1.8245 - categorical_accuracy: 0.3499 - val_loss: 1.7338 - val_categorical_accuracy: 0.1600 - lr: 0.0010
Epoch 2/2
11/12 [==========================>...] - ETA: 3s - loss: 1.2210 - categorical_accuracy: 0.5227
Epoch 00002: saving model to model_init_2021-03-1511_38_25.032126/model-00002-1.22527-0.52187-2.99775-0.17000.h5
12/12 [==============================] - 48s 4s/step - loss: 1.2253 - categorical_accuracy: 0.5219 - val_loss: 2.9977 - val_categorical_accuracy: 0.1700 - lr: 0.0010


In [17]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=80,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:40:10.377314 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
9/9 [==============================] - ETA: 0s - loss: 1.7715 - categorical_accuracy: 0.3409
Epoch 00001: saving model to model_init_2021-03-1511_40_10.376694/model-00001-1.77154-0.34087-1.67701-0.16000.h5
9/9 [==============================] - 48s 5s/step - loss: 1.7715 - categorical_accuracy: 0.3409 - val_loss: 1.6770 - val_categorical_accuracy: 0.1600 - lr: 0.0010
Epoch 2/2
9/9 [==============================] - ETA: 0s - loss: 1.0863 - categorical_accuracy: 0.5716
Epoch 00002: saving model to model_init_2021-03-1511_40_10.376694/model-00002-1.08630-0.57164-2.28023-0.22000.h5
9/9 [==============================] - 45s 5s/step - loss: 1.0863 - categorical_accuracy: 0.5716 - val_loss: 2.2802 - val_categorical_accuracy: 0.2200 - lr: 0.0010


In [18]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=160,image_width=160)
conv_3d1.initialize_hyperparams(frames_to_sample=30,batch_size=15,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:41:53.864535 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 1736389
Epoch 1/2
45/45 [==============================] - ETA: 0s - loss: 1.5728 - categorical_accuracy: 0.3982
Epoch 00001: saving model to model_init_2021-03-1511_41_53.863774/model-00001-1.57278-0.39819-2.89440-0.24000.h5
45/45 [==============================] - 109s 2s/step - loss: 1.5728 - categorical_accuracy: 0.3982 - val_loss: 2.8944 - val_categorical_accuracy: 0.2400 - lr: 0.0010
Epoch 2/2
45/45 [==============================] - ETA: 0s - loss: 1.2310 - categorical_accuracy: 0.5083
Epoch 00002: saving model to model_init_2021-03-1511_41_53.863774/model-00002-1.23101-0.50830-4.64528-0.24000.h5
45/45 [==============================] - 114s 3s/step - loss: 1.2310 - categorical_accuracy: 0.5083 - val_loss: 4.6453 - val_categorical_accuracy: 0.2400 - lr: 0.0010


In [19]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=160,image_width=160)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=15,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:45:45.731355 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 1736389
Epoch 1/2
45/45 [==============================] - ETA: 0s - loss: 1.6535 - categorical_accuracy: 0.4133
Epoch 00001: saving model to model_init_2021-03-1511_45_45.716693/model-00001-1.65351-0.41327-5.78935-0.21000.h5
45/45 [==============================] - 95s 2s/step - loss: 1.6535 - categorical_accuracy: 0.4133 - val_loss: 5.7893 - val_categorical_accuracy: 0.2100 - lr: 0.0010
Epoch 2/2
45/45 [==============================] - ETA: 0s - loss: 1.0873 - categorical_accuracy: 0.5958
Epoch 00002: saving model to model_init_2021-03-1511_45_45.716693/model-00002-1.08731-0.59578-8.10266-0.21000.h5
45/45 [==============================] - 61s 1s/step - loss: 1.0873 - categorical_accuracy: 0.5958 - val_loss: 8.1027 - val_categorical_accuracy: 0.2100 - lr: 0.0010


In [20]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=15,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:48:28.521665 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
44/45 [============================>.] - ETA: 0s - loss: 1.6826 - categorical_accuracy: 0.3818
Epoch 00001: saving model to model_init_2021-03-1511_48_28.520854/model-00001-1.68579-0.38160-3.02213-0.16000.h5
45/45 [==============================] - 47s 1s/step - loss: 1.6858 - categorical_accuracy: 0.3816 - val_loss: 3.0221 - val_categorical_accuracy: 0.1600 - lr: 0.0010
Epoch 2/2
45/45 [==============================] - ETA: 0s - loss: 1.1809 - categorical_accuracy: 0.5294
Epoch 00002: saving model to model_init_2021-03-1511_48_28.520854/model-00002-1.18086-0.52941-4.54284-0.13000.h5
45/45 [==============================] - 47s 1s/step - loss: 1.1809 - categorical_accuracy: 0.5294 - val_loss: 4.5428 - val_categorical_accuracy: 0.1300 - lr: 0.0010


In [21]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=10,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:50:06.785456 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
66/67 [============================>.] - ETA: 0s - loss: 1.7452 - categorical_accuracy: 0.3545
Epoch 00001: saving model to model_init_2021-03-1511_50_06.784848/model-00001-1.74835-0.35445-2.21104-0.19000.h5
67/67 [==============================] - 48s 709ms/step - loss: 1.7484 - categorical_accuracy: 0.3544 - val_loss: 2.2110 - val_categorical_accuracy: 0.1900 - lr: 0.0010
Epoch 2/2
66/67 [============================>.] - ETA: 0s - loss: 1.1390 - categorical_accuracy: 0.5561
Epoch 00002: saving model to model_init_2021-03-1511_50_06.784848/model-00002-1.14340-0.55505-3.64934-0.21000.h5
67/67 [==============================] - 46s 693ms/step - loss: 1.1434 - categorical_accuracy: 0.5551 - val_loss: 3.6493 - val_categorical_accuracy: 0.2100 - lr: 0.0010


In [22]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=100,image_width=100)
conv_3d1.initialize_hyperparams(frames_to_sample=30,batch_size=10,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:51:45.202952 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 687813
Epoch 1/2
67/67 [==============================] - ETA: 0s - loss: 1.6941 - categorical_accuracy: 0.3605
Epoch 00001: saving model to model_init_2021-03-1511_51_45.202327/model-00001-1.69413-0.36048-2.38097-0.20000.h5
67/67 [==============================] - 183s 3s/step - loss: 1.6941 - categorical_accuracy: 0.3605 - val_loss: 2.3810 - val_categorical_accuracy: 0.2000 - lr: 0.0010
Epoch 2/2
67/67 [==============================] - ETA: 0s - loss: 1.4527 - categorical_accuracy: 0.4480
Epoch 00002: saving model to model_init_2021-03-1511_51_45.202327/model-00002-1.45271-0.44796-2.98445-0.15000.h5
67/67 [==============================] - 187s 3s/step - loss: 1.4527 - categorical_accuracy: 0.4480 - val_loss: 2.9845 - val_categorical_accuracy: 0.1500 - lr: 0.0010


In [ ]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=160,image_width=160)
conv_3d1.initialize_hyperparams(frames_to_sample=30,batch_size=10,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

W0315 11:58:01.496024 140545046239040 callbacks.py:1071] `period` argument is deprecated. Please use `save_freq` to specify the frequency in number of batches seen.


Total Params: 1736389
Epoch 1/2
67/67 [==============================] - ETA: 0s - loss: 1.5371 - categorical_accuracy: 0.4329

In [ ]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=160,image_width=160)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=10,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

In [ ]:
conv_3d1=ModelConv3D1()
conv_3d1.initialize_src_path(main_folder)
conv_3d1.initialize_image_properties(image_height=160,image_width=160)
conv_3d1.initialize_hyperparams(frames_to_sample=16,batch_size=40,num_epochs=2)
conv_3d1_model=conv_3d1.define_model()
print("Total Params:", conv_3d1_model.count_params())
conv_3d1.train_model(conv_3d1_model)

### Observation

**As we see from the above experiments image resolution and number of frames in sequence have more impact on training time than batch_size.**

So experimentations are carried with batch size fixed around 15-40 and changing the resolution and number of image per sequence based on the device memory constraints . Models are designed such that their memory foot print is less than 50 MB which corresponds to 4.3 million parameters assuming the datatype size of parameters to be 12 bytes